tests/test_client_functional.py

```diff
@@ -2927,6 +2927,73 @@ async def test_creds_in_auth_and_url() -> None:
         await session.close()
 
 
+@pytest.mark.parametrize(
+    ["check_redirect_auth_precedence"],
+    ([False], [True]),
+    ids=("redirect auth clash allowed", "redirect auth precedence"),
+)
+async def test_creds_in_auth_and_redirect_url(
+    create_server_for_url_and_handler: Callable[[URL, Handler], Awaitable[TestServer]],
+    check_redirect_auth_precedence: bool,
+) -> None:
+    url_from = URL("http://example.com")
+    url_to = URL("http://user@example.com")
+    redirected = False
+
+    async def srv(request: web.Request) -> web.Response:
+        nonlocal redirected
+
+        assert request.host == url_from.host
+
+        if not redirected:
+            redirected = True
+            raise web.HTTPMovedPermanently(url_to)
+
+        return web.Response()
+
+    server = await create_server_for_url_and_handler(url_from, srv)
+
+    etc_hosts = {
+        (url_from.host, 80): server,
+    }
+
+    class FakeResolver(AbstractResolver):
+        async def resolve(
+            self,
+            host: str,
+            port: int = 0,
+            family: socket.AddressFamily = socket.AF_INET,
+        ) -> List[ResolveResult]:
+            server = etc_hosts[(host, port)]
+            assert server.port is not None
+
+            return [
+                {
+                    "hostname": host,
+                    "host": server.host,
+                    "port": server.port,
+                    "family": socket.AF_INET,
+                    "proto": 0,
+                    "flags": socket.AI_NUMERICHOST,
+                }
+            ]
+
+        async def close(self) -> None:
+            """Dummy"""
+
+    connector = aiohttp.TCPConnector(resolver=FakeResolver(), ssl=False)
+
+    async with aiohttp.ClientSession(connector=connector) as client, client.get(
+        url_from, auth=aiohttp.BasicAuth("user", "pass")
+    ) as resp:
+        assert len(resp.history) == 1
+        assert str(resp.url) == "http://example.com"
+        assert resp.status == 200
+
+        if check_redirect_auth_precedence:
+            assert resp.request_info.headers.get("authorization") == "Basic dXNlcjo="
```

Comment:
This should probably have an `else` branch with the opposite assertion.

Reply:
I suppose this `else` branch doesn't make sense anymore in light of our discussion below?

Reply:
I'm not entirely sure. I'd like to hear what @Dreamsorcerer and @bdraco have to say.

Reply:
Wouldn't this always be true? Seems to me like the parametrize is just deciding whether to check this or not. In which case, wouldn't it make more sense to just add a comment on this assert explaining the likely cause, rather than running the test twice with parametrize?

Reply:
That's a good point actually, I'll do that instead.